Implement Subresource Integrity

nmvk opened the pull request:
Implemented response validation part of https://w3c.github.io/webappsec-subresource-integrity/.
Implemented step eighteen of the main fetch. If a request has integrity metadata, then following steps are performed
1) Wait for response body
2) If the response does not have a termination reason and response does not match request’s integrity metadata, set response and internalResponse to a network error.

Dependency updated: html5ever-atoms from 0.1.2 to 0.1.3. This will not completely fix #14523, It will implement changes related to response validation. Request validation algorithm implementation needs CSP.

I did not update any WPT-Test. In my local system, I found some assertion issue dependent on the order of execution of test-case. It would be helpful if someone could do "try" build on these changes to get wpt results.

r? @jdm 
<!-- Please describe your changes on the following line: -->


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors

<!-- Either: -->
- [X] There are tests for these changes 


<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/14865)
<!-- Reviewable:end -->

mrnayak commented:
r? @jdm

jdm commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit a64b964 with merge 2457bcf...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/1749)

jdm commented:
```
  ▶ TIMEOUT [expected OK] /subresource-integrity/subresource-integrity.sub.html
  │ 
  │ VMware, Inc.
  │ Gallium 0.4 on softpipe
  └ 3.3 (Core Profile) Mesa 12.0.1

  ▶ Unexpected subtest result in /subresource-integrity/subresource-integrity.sub.html:
  └ NOTRUN [expected PASS] Style: Same-origin with correct sha256 and sha512 hash, rel='alternate stylesheet' enabled

  ▶ Unexpected subtest result in /subresource-integrity/subresource-integrity.sub.html:
  └ NOTRUN [expected PASS] Style: Same-origin with incorrect sha256 and sha512 hash, rel='alternate stylesheet' enabled

  ▶ Unexpected subtest result in /subresource-integrity/subresource-integrity.sub.html:
  │ FAIL [expected PASS] Style: <crossorigin='anonymous'> with correct hash, ACAO: *
  │   → assert_unreached: Good load fired error handler. Reached unreachable code
  │ FAIL [expected PASS] Style: <crossorigin='use-credentials'> with correct hash, CORS-eligible
  │   → assert_unreached: Good load fired error handler. Reached unreachable code
  │ 
  │ SRIStyleTest.prototype.execute/</<@http://web-platform.test:8000/subresource-integrity/subresource-integrity.sub.html:132:39
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  └ SRIStyleTest.prototype.execute/<@http://web-platform.test:8000/subresource-integrity/subresource-integrity.sub.html:132:17

  ▶ Unexpected subtest result in /subresource-integrity/subresource-integrity.sub.html:
  │ FAIL [expected PASS] Script: <crossorigin='use-credentials'> with correct hash, CORS-eligible
  │   → assert_unreached: Good load fired error handler. Reached unreachable code
  │ 
  │ SRIScriptTest.prototype.execute/</<@http://web-platform.test:8000/subresource-integrity/subresource-integrity.sub.html:59:39
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  └ SRIScriptTest.prototype.execute/<@http://web-platform.test:8000/subresource-integrity/subresource-integrity.sub.html:59:17

  ▶ Unexpected subtest result in /subresource-integrity/subresource-integrity.sub.html:
  │ FAIL [expected PASS] Style: Same-origin with incorrect hash.
  │   → assert_not_equals: got disallowed value "rgb(255, 255, 0)"
  │ FAIL [expected PASS] Style: Same-origin with sha256 match, sha512 mismatch
  │   → assert_not_equals: got disallowed value "rgb(255, 255, 0)"
  │ FAIL [expected PASS] Style: <crossorigin='anonymous'> with incorrect hash, ACAO: *
  │   → assert_not_equals: got disallowed value "rgb(255, 255, 0)"
  │ FAIL [expected PASS] Style: <crossorigin='use-credentials'> with incorrect hash CORS-eligible
  │   → assert_not_equals: got disallowed value "rgb(255, 255, 0)"
  │ FAIL [expected PASS] Style: <crossorigin='anonymous'> with CORS-ineligible resource
  │   → assert_not_equals: got disallowed value "rgb(255, 255, 0)"
  │ FAIL [expected PASS] Style: Cross-origin, not CORS request, with correct hash
  │   → assert_not_equals: got disallowed value "rgb(255, 255, 0)"
  │ FAIL [expected PASS] Style: Cross-origin, not CORS request, with hash mismatch
  │   → assert_not_equals: got disallowed value "rgb(255, 255, 0)"
  │ 
  │ SRIStyleTest.prototype.execute/</<@http://web-platform.test:8000/subresource-integrity/subresource-integrity.sub.html:141:21
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  └ SRIStyleTest.prototype.execute/<@http://web-platform.test:8000/subresource-integrity/subresource-integrity.sub.html:139:17
```

jdm commented:
Let me know if you want me to start reviewing the changes or whether I should wait until you look into the test failure.

mrnayak commented:
Thanks Josh, I will work on these test cases today. Will notify you once these gets fixed.

mrnayak commented:
None of the test case failures seem to be related to changes.

1) assertion error in 7 test case is due to stylesheet already been loaded in other success test case. I moved all test case expecting assert_not_equals to execute before any success test case.
2) crossorigin has not been implemented in htmllinkelement. So i changed assertion of two test case (`Style: <crossorigin='anonymous'> with correct hash, ACAO: *` and `Style: <crossorigin='use-credentials'> with correct hash, CORS-eligible`) to false.
3) Two alternate stylesheet related tests got timed out since we ignore alternate stylesheet. I have filed an issue #14881. Currently, I have commented these tests.
4) `Script: <crossorigin='use-credentials'> with correct hash, CORS-eligible` failed due to bug in step six of cors_check of `http_loader` . This has been fixed.

@jdm If my above analysis is correct, can you please review the changes.

jdm commented:
Rather than modifying the test file, we should be modifying `tests/wpt/metadata/subresource-integrity/subresource-integrity.sub.html.ini` to include the expected failing test results.

jdm commented:
Can you explain point 1 in more detail? Why does the order make a difference in the test result?

mrnayak commented:
@jdm I am not sure if point 1 is a bug. Our assertion is to validate that div style background color is not yellow.  Moment first success test case executes stylesheet gets loaded and subsequent all assertion fail though we don't load stylesheet again. 

I update `tests/wpt/metadata/subresource-integrity/subresource-integrity.sub.html.ini`, Thanks

jdm commented:
I know what the problem is - the test relies on removing the stylesheet from any previous test before starting the new test. However, Servo doesn't support removing stylesheets from the document yet (see #976).

mrnayak commented:
So I will modify `subresource-integrity.sub.html.ini` to include these tests as well?

jdm commented:
Yes please!

mrnayak commented:
@jdm I have done the changes

jdm commented on tests/unit/net/fetch.rs:
```diff
@@ -556,6 +555,57 @@ fn test_fetch_with_hsts() {
+
+    let _ = server.close();
+    assert!(response.is_network_error());
+    assert!(response.internal_response.unwrap().is_network_error());
```
This assertion won't pass after previous changes I requested. We should be able to remove it.

jdm commented on tests/unit/net/fetch.rs:
```diff
@@ -556,6 +555,57 @@ fn test_fetch_with_hsts() {
+        let mut request = Request::new(url, Some(origin), false, None);
+        *request.referrer.borrow_mut() = Referrer::NoReferrer;
+        *request.integrity_metadata.borrow_mut()
+             = "sha384-H8BRh8j48O9oYatfu5AZzq6A9RINhZO5H16dQZngK7T62em8MUt1FLm52t+eX6xO".to_owned();
```
Let's add a comment about how this was generated.

jdm commented on tests/unit/net/fetch.rs:
```diff
@@ -556,6 +555,57 @@ fn test_fetch_with_hsts() {
+    };
+    let (mut server, server_url) = make_server(handler);
+
+    let do_fetch = |url: ServoUrl| {
```
This code can be inline instead of using a closure.

jdm commented on tests/unit/net/subresource_integrity.rs:
```diff
@@ -0,0 +1,84 @@
+}
+
+#[test]
+fn test_parsed_metadata_without_options() {
```
Let's add a test for malformed integrity attributes, too. No `-`s, multiple `?`s, etc.

jdm commented on components/net/subresource_integrity.rs:
```diff
@@ -0,0 +1,164 @@
+use std::sync::MutexGuard;
+lazy_static! {
+    //Key is supported algorithm and value is priority
+    static ref SUPPORTED_ALGORITHM: HashMap<&'static str, usize> = {
```
We can do this instead:
```rust
const SUPPORTED_ALGORITHM: &'static [&'static str] = [
    "sha256",
    "sha384",
    "sha512",
];
```
and use the position in the array as the priority instead.

jdm commented on components/net/subresource_integrity.rs:
```diff
@@ -0,0 +1,164 @@
+/// https://w3c.github.io/webappsec-subresource-integrity/#getprioritizedhashfunction
+pub fn get_prioritized_hash_function(hash_func_left: &str, hash_func_right: &str) -> String {
+    let left_priority = SUPPORTED_ALGORITHM.get(hash_func_left).unwrap();
+    let right_priority = SUPPORTED_ALGORITHM.get(hash_func_right).unwrap();
```
These can use `SUPPORTED_ALGORITHM.position(|s| s == hash_func_left).unwrap()` instead.

jdm commented on components/net/subresource_integrity.rs:
```diff
@@ -0,0 +1,164 @@
+}
+
+/// https://w3c.github.io/webappsec-subresource-integrity/#getprioritizedhashfunction
+pub fn get_prioritized_hash_function(hash_func_left: &str, hash_func_right: &str) -> String {
```
Let's return `Option<&str>` here instead, which will avoid the magic empty string behaviour.

jdm commented on components/net/subresource_integrity.rs:
```diff
@@ -0,0 +1,164 @@
+    let mut result: Vec<SriEntry> = vec![integrity_metadata_list[0].clone()];
+    let mut current_algorithm = result[0].alg.clone();
+
+    for i in 1..integrity_metadata_list.len() {
```
```rust
for integrity_metadata in &integrity_metadata_list[1..] {
```

jdm commented on components/net/subresource_integrity.rs:
```diff
@@ -0,0 +1,164 @@
+        let intgerity_metadata = integrity_metadata_list[i].clone();
+
+        let prioritized_hash = get_prioritized_hash_function(&intgerity_metadata.alg,
+                                                                 &*current_algorithm);
```
nit: indentation.

jdm commented on components/net/subresource_integrity.rs:
```diff
@@ -0,0 +1,164 @@
+        let response_digest = hash(message_digest, vec);
+        response_digest.to_base64(STANDARD)
+    } else {
+        "".to_owned()
```
We should probably make this `unreachable!("Tried to calculate digest of incomplete response body")` instead.

jdm commented on components/net/subresource_integrity.rs:
```diff
@@ -0,0 +1,164 @@
+    let parsed_metadata_list: Vec<SriEntry> = parsed_metadata(integrity_metadata);
+
+    // Step 2 & 4
+    if parsed_metadata_list.len() == 0 {
```
`if parse_metadata_list.is_empty() {`

jdm reviewed (changes requested):
This is really great work! I have a number of small changes, but in general I like the way the code is organized and found the changes easy to read.

mrnayak commented:
Thanks @jdm I have made changes. Please let me know if there are any other changes.

jdm commented:
@bors-servo: r+
Wonderful!

bors-servo commented:
:pushpin: Commit 49aff3e has been approved by `jdm`

<!-- @bors-servo r=jdm 49aff3e58efb9a386dbb8dadcc15b9b16701230c -->

bors-servo commented:
:hourglass: Testing commit 49aff3e with merge f3fe099...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/1761)

mrnayak commented:
I had missed updating expectation of two test case :(. I have corrected this.

jdm commented:
You will also need to adjust the expectations for /html/dom/interfaces.html, /html/dom/reflection-misc.html, and /html/dom/reflection-metadata.html. I recommend running
`./mach test-wpt tests/wpt/web-platform-test/path/to/test.html --no-pause-after-test --log-raw /tmp/servo.log && ./mach update-wpt /tmp/servo.log` for each of those tests.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #14867) made this pull request unmergeable. Please resolve the merge conflicts.

nmvk committed:
Implement Subresource Integrity

Implemented response validation part of
https://w3c.github.io/webappsec-subresource-integrity/.
Implemented step eighteen of the main fetch. If a request has integrity
metadata, then following steps are performed
*Wait for response body
*If the response does not have a termination reason and response does not
match request’s integrity metadata, set response to a
network error.# Please enter the commit message for your changes. Lines starting

mrnayak commented:
@jdm Thanks, I have fixed expectations and merge conflicts

jdm commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit a302649 has been approved by `jdm`

<!-- @bors-servo r=jdm a3026499f43860c0d02170534bee20f8f5cc7faa -->

bors-servo commented:
:hourglass: Testing commit a302649 with merge f958daf...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/5094)

mrnayak commented:
Infra issue?

KiChjang commented:
@bors-servo retry

bors-servo commented:
:zap: Previous build results for [android](http://build.servo.org/builders/android/builds/4794), [arm64](http://build.servo.org/builders/arm64/builds/5000), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/1778), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/5069), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/1775), [windows-gnu-dev](http://build.servo.org/builders/windows-gnu-dev/builds/233), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/234) are reusable. Rebuilding only [arm32](http://build.servo.org/builders/arm32/builds/4960), [linux-dev](http://build.servo.org/builders/linux-dev/builds/5094), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/1771), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/5110), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/1814)...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/4794), [arm32](http://build.servo.org/builders/arm32/builds/4960), [arm64](http://build.servo.org/builders/arm64/builds/5000), [linux-dev](http://build.servo.org/builders/linux-dev/builds/5095), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/1771), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/1778), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/5069), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/5110), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/1775), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/1814), [windows-gnu-dev](http://build.servo.org/builders/windows-gnu-dev/builds/233), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/234)
